src/api/providers/gemini.ts

```diff
@@ -16,27 +17,59 @@ import { BaseProvider } from "./base-provider"
 export class GeminiHandler extends BaseProvider implements SingleCompletionHandler {
 	protected options: ApiHandlerOptions
 	private client: GoogleGenAI
+	private contentCaches: Map<string, { key: string; count: number }>
 
 	constructor(options: ApiHandlerOptions) {
 		super()
 		this.options = options
 		this.client = new GoogleGenAI({ apiKey: options.geminiApiKey ?? "not-provided" })
+		this.contentCaches = new Map()
 	}
 
-	async *createMessage(systemPrompt: string, messages: Anthropic.Messages.MessageParam[]): ApiStream {
-		const { id: model, thinkingConfig, maxOutputTokens } = this.getModel()
+	async *createMessage(
+		systemInstruction: string,
+		messages: Anthropic.Messages.MessageParam[],
+		taskId?: string,
+	): ApiStream {
+		const { id: model, thinkingConfig, maxOutputTokens, supportsPromptCache } = this.getModel()
+
+		const contents = messages.map(convertAnthropicMessageToGemini)
+		let uncachedContent: Content[] | undefined = undefined
+		let cachedContent: string | undefined = undefined
+		let cacheWriteTokens: number = 0
+
+		// https://ai.google.dev/gemini-api/docs/caching?lang=node
+		if (supportsPromptCache && taskId) {
+			const cacheEntry = this.contentCaches.get(taskId)
+
+			if (cacheEntry) {
+				uncachedContent = contents.slice(cacheEntry.count, contents.length)
+				cachedContent = cacheEntry.key
+			}
+
+			const newCacheEntry = await this.client.caches.create({
+				model,
+				config: { contents, systemInstruction, ttl: "300s" },
+			})
+
+			if (newCacheEntry.name) {
+				this.contentCaches.set(taskId, { key: newCacheEntry.name, count: contents.length })
```

Comment:
Should we also try to expire the old cache at this point?

Reply:
Yeah, probably should.

Reply:
Also, I think the index might be invalid on a checkpoint restore. I think we might want the cline instance to maintain the cache key so it can bust it as needed.